examples/structural_mechanics/crash/train.py

```diff
@@ -0,0 +1,271 @@
+            pred = self.model(sample=sample, data_stats=self.data_stats)
+
+            # Reshape target
+            target_flat = sample.node_target  # [N, T*Fo]
+            N = target_flat.size(0)
+            Fo = 3  # output features per node
+            assert target_flat.size(1) == T * Fo, (
+                f"target dim {target_flat.size(1)} != {T * Fo}"
+            )
+            target = target_flat.view(N, T, Fo).transpose(0, 1).contiguous()  # [T,N,Fo]
+
+            return self.criterion(pred, target)
+
+    def backward(self, loss):
+        if self.amp:
+            self.scaler.scale(loss).backward()
+            self.scaler.step(self.optimizer)
+            self.scaler.update()
+        else:
+            loss.backward()
+            self.optimizer.step()
+
+
+@hydra.main(version_base="1.3", config_path="conf", config_name="config")
+def main(cfg: DictConfig) -> None:
+    DistributedManager.initialize()
+    dist = DistributedManager()
+
+    logger = PythonLogger("main")
+    logger0 = RankZeroLoggingWrapper(logger, dist)
+    logger0.file_logging()
+
+    trainer = Trainer(cfg, logger0)
+    logger0.info("Training started...")
+
+    for epoch in range(trainer.epoch_init, cfg.training.epochs):
+        if trainer.sampler is not None:
+            trainer.sampler.set_epoch(epoch)
+
+        progress_bar = tqdm(
+            trainer.dataloader,
+            desc=f"Epoch {epoch + 1}/{cfg.training.epochs}",
+            leave=False,
+            disable=True,
+        )
+
+        total_loss = 0.0
+        num_batches = 0
+        start = time.time()
+
+        for sample in progress_bar:
+            sample = sample[0].to(dist.device)  # SimSample .to()
+            loss = trainer.train(sample, epoch)
+            total_loss += loss.item()
+            num_batches += 1
+
+            progress_bar.set_postfix(loss=f"{loss.item():.3e}")
+            del sample
+            torch.cuda.empty_cache()
+
+        trainer.scheduler.step()
+
+        avg_loss = total_loss / max(num_batches, 1)
+        logger0.info(
+            f"epoch: {epoch + 1}, avg_loss: {avg_loss:10.3e}, "
+            f"lr: {trainer.optimizer.param_groups[0]['lr']:.3e}, "
+            f"time per epoch: {(time.time() - start):10.3e}"
+        )
+
+        if dist.rank == 0:
+            trainer.writer.add_scalar("loss", avg_loss, epoch)
+            trainer.writer.add_scalar(
+                "learning_rate", trainer.optimizer.param_groups[0]["lr"], epoch
+            )
+
+        if dist.world_size > 1:
+            torch.distributed.barrier()
+        if dist.rank == 0:
+            save_checkpoint(
+                cfg.training.ckpt_path,
+                models=trainer.model,
+                optimizer=trainer.optimizer,
+                scheduler=trainer.scheduler,
+                scaler=trainer.scaler,
+                epoch=epoch + 1,
+            )
+            logger.info(f"Saved model on rank {dist.rank}")
+
+        torch.cuda.empty_cache()
```

Comment:
Is this needed since you already have it in the epoch loop?
Maybe keep this and remove the other one? Or you get OOM in this case?